File: services/saas-api/src/saas_api/schemas/reports.py
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from saas_api.schemas.birth_profile import BirthProfileInput


class FreeReportRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    tenant_slug: str = Field(alias="tenantSlug")
    birth_profile: BirthProfileInput = Field(alias="birthProfile")
    locale: str = "en"


class ReportListItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    type: str
    title: str
    generated_at: str = Field(alias="generatedAt")


class ReportStatusResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    status: str
    report_type: str = Field(alias="reportType")
    locale: str
    error_code: str | None = Field(default=None, alias="errorCode")
    error_message: str | None = Field(default=None, alias="errorMessage")
    created_at: str = Field(alias="createdAt")
    completed_at: str | None = Field(default=None, alias="completedAt")
    report: dict[str, Any] | None = None

File: packages/api-contracts/src/reports.ts
import type { BirthProfile, BirthProfileTopic, Report } from "@astro/tenant-config";
import { z } from "zod";

export interface FreeReportRequest {
  tenantSlug?: string;
  tenantId?: string;
  birthProfile?: BirthProfile;
  topic?: BirthProfileTopic;
  userId?: string;
}

export interface ReportListItem {
  id: string;
  type: Report["type"];
  title: string;
  generatedAt: string;
}

export interface ListReportsQuery {
  tenantId?: string;
  userId?: string;
}

export type { Report };

export type ReportLifecycleStatus = "pending" | "generating" | "completed" | "failed";

export interface ReportStatusResponse {
  id: string;
  status: ReportLifecycleStatus;
  reportType: string;
  locale: string;
  errorCode?: string | null;
  errorMessage?: string | null;
  createdAt: string;
  completedAt?: string | null;
  report?: Report | null;
}

export const reportListItemSchema = z.object({
  id: z.string().min(1),
  type: z.enum(["free", "natal", "compatibility", "forecast", "custom"]),
  title: z.string().min(1),
  generatedAt: z.string(),
});

export const reportListSchema = z.array(reportListItemSchema);

export const reportStatusResponseSchema = z.object({
  id: z.string().min(1),
  status: z.enum(["pending", "generating", "completed", "failed"]),
  reportType: z.string().min(1),
  locale: z.string().min(1),
  errorCode: z.string().nullable().optional(),
  errorMessage: z.string().nullable().optional(),
  createdAt: z.string(),
  completedAt: z.string().nullable().optional(),
  report: z.unknown().nullable().optional(),
});
